Add BackpackBlock

gr8pefish opened the pull request:
First pass:

Added /block package, containing AbstractBlockTE, BlockBackpack, and TEBackpack
Altered ContainerBackpack to have an overloaded constructor for the TE as well as the ItemStack
Added (untested) IronBackpacksAPI.applyPackInfo(tile)
Added (untested) BackpackInfo.fromTE(tile)
Minor changes (renamed creative_tab, and renamed inventory_capability_handler)

gr8pefish committed:
First stage, *very* incomplete

Added /block package, containing AbstractBlockTE, BlockBackpack, and TEBackpack
Altered ContainerBackpack to have an overloaded constructor for the TE as well as the ItemStack
Added (untested) IronBackpacksAPI.applyPackInfo(tile)
Added (untested) BackpackInfo.fromTE(tile)
Minor changes (renamed creative_tab, and renamed inventory_capability_handler)

gr8pefish commented:
From the chat:

Alright, so bear with me, here was my general idea:

1) R-click the ItemStack (of and ItemBackpack)
2) Makes a BlockBackpack of the appropriate BackpackVariant
3) That block (specifically the BackpackTileEntity) has the BackpackInfo taken from the (now-deleted) ItemStack
4) The TE therefore uses the IItemHandler, same serialization, etc. as the ItemStack. Easy. Life is full of rainbows and sunshine
5) When the block is broken it takes the BackpackInfo from the TE to drop an ItemBackpack (with that info copied back to it) into the world.

Obviously, that was not reality. I have to: 
* Add in the r-click code for item stack (actually mostly wirtten, but have to copy over from my TestMod)
* Add registration for all variants of the Block (or just the single block with the Variant as a param in the constructor)
* Initialize and access the TE data correctly (big one)
* Figure out the black magic that is ItemStack IItemHandler capabilities so I can use the same one for the TE
* Rewrite lots of the methods (that are currently hardcoded for itemstacks) so that #4 above can actually exist in reality.
* Do the custom dropping of an ItemBackpack from the TE's data.
* Copy over the models/blockstate JSONs/textures (from the TestMod)

TehNut commented on src/main/java/gr8pefish/ironbackpacks/IronBackpacks.java:
```diff
@@ -25,7 +25,7 @@
     public static final Logger LOGGER = LogManager.getLogger(NAME);
     public static final SimpleNetworkWrapper NETWORK = new SimpleNetworkWrapper(MODID);
-    public static final CreativeTabs TAB_IB = new CreativeTabs(MODID) {
+    public static final CreativeTabs CREATIVE_TAB_IB = new CreativeTabs(MODID) {
```
whytho

gr8pefish replied:
Minor, but more clear. Shouldn't have included it in this commit, but IronBackpacks.CREATIVE_TAB_IB makes more sense than just TAB_IB.

TehNut replied:
While I agree field names should be descriptive, there's also this thing called an IDE that helps you write code. 
![image](https://user-images.githubusercontent.com/5789439/35467565-cc97de02-02c4-11e8-882d-b7b065fcdc0f.png)

gr8pefish replied:
I'll defer to you on this, no biggie. Although, to be fair, the IDE does help autocomplete, so long names are no big deal.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/api/IronBackpacksAPI.java:
```diff
@@ -198,6 +199,35 @@ public static ItemStack applyPackInfo(@Nonnull ItemStack stack, @Nonnull Backpac
+     * @return - The updated tileEntity with the info applied.
+     */
+    @Nonnull
+    public static TileEntity applyPackInfo(@Nonnull TileEntity tileEntity, @Nonnull BackpackInfo backpackInfo) {
```
We should revisit these application methods to condense down. Also, there's no need for this to be in the API. Mods shouldn't be creating their own tiles.

gr8pefish replied:
Yup, good point.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/api/IronBackpacksAPI.java:
```diff
@@ -198,6 +199,35 @@ public static ItemStack applyPackInfo(@Nonnull ItemStack stack, @Nonnull Backpac
+        Preconditions.checkNotNull(tileEntity, "TileEntity cannot be null");
+        Preconditions.checkNotNull(backpackInfo, "BackpackInfo cannot be null");
+
+        NBTTagCompound tagCompound = tileEntity.getTileData();
```
Don't modify the NBT directly. Set the fields. When the TE saves, it'll serialize as needed.

gr8pefish replied:
~~So you'd do `BackpackTE.getMyField() = 5` or similar then?~~

`BackpackTileEntity tile = RegistrarIronBackpacks.TILE;`
`BackpackInfo info = tile.getBackpackInfo();`
`//modify info here`

TehNut replied:
Setters would be used, yes.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/api/backpack/BackpackInfo.java:
```diff
@@ -282,11 +283,33 @@ public static BackpackInfo fromStack(@Nonnull ItemStack stack) {
 
+    //TODO: Test
+    @Nonnull
+    public static BackpackInfo fromTileEntity(@Nonnull TileEntity te) {
```
Again, shouldn't be in the API. If it *must* be, then it should be a cap. ~~Maybe we can genericize the stack one?~~ You did that it seems.

gr8pefish replied:
Started doing that. Have to figure out exactly how the cap works again, but that's the idea.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/AbstractBlockTileEntity.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @param <TE> - {@link TileEntity}
+ */
+public abstract class AbstractBlockTileEntity<TE extends TileEntity> extends Block {
```
I vote for just removing this class. It's useful in larger mods with many tiles, but it's just a waste of precious hard drive space. All 1.58Kb of it.

gr8pefish replied:
Sure haha, it can be removed. I was just doing it for my own sanity, as I didn't want a giant file to mess with to start with.

Edit: Will do, but later

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+    public static final PropertyDirection FACING = BlockHorizontal.FACING;
+
+    /** Constructor */
+    public BlockBackpack(BackpackVariant backpackVariant) {
```
I'll be unflattening this for 1.12. I have an idea. 👍

gr8pefish replied:
Okay, I assume I'll just let you handle that then.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+    @Nullable
+    @Override
+    public TileEntityBackpack createTileEntity(World world, IBlockState state) {
+        return new TileEntityBackpack(placeholder); //TODO: BackpackInfo from the itemStack
```
This should use a default constructor. When we set the block, we will modify the tile to include the data we need.

gr8pefish replied:
To clarify: So the default constructor with no data, then on the block's placement we include the BackpackInfo?

TehNut replied:
Yes. On block placement it would look something like

```java
world.setBlockState(pos, RegistrarIronBackpacks.BACKPACK_BLOCK);
TileEntityBackpack tileBackpack = new TileEntityBackpack();
tileBackpack.setBackpackInfo(backpackInfo);
world.setTileEntity(pos, tileBackpack);
```

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
+        if (!world.isRemote) {
+            player.openGui(IronBackpacks.INSTANCE, GuiHandler.OPEN_GUI_TE_ID, world, pos.getX(), pos.getY(), pos.getZ()); //ToDo: Open correct GUI
+            getTileEntity(world, pos).markDirty(); //necessary?
```
Should not be necessary. `markDirty()` tells the world the data has changed and should be saved during the next save cycle.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+
+    /** When breaking the block, include the contents of the inventory. */
+    @Override
+    public void breakBlock(World world, BlockPos pos, IBlockState state) {
```
Maybe just drop the backpack item? I can see many players finding this a major annoyance.

gr8pefish replied:
Yeah, that's a better idea, good call 👍

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+        //Also drop itself
+        this.dropBlockAsItem(world, pos, state, 0);
+        //necessary?
+        super.breakBlock(world, pos, state);
```
Very necessary unless you remove the tile yourself.

gr8pefish replied:
Okay, thanks for the clarification

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+
+    @Override
+    public IBlockState getStateFromMeta(int meta) {
+        EnumFacing facing = EnumFacing.Plane.HORIZONTAL.facings()[meta & 0b11]; //first 2 bits for facing, 4 possibilities
```
```java
return getDefaultState().withValue(FACING, EnumFacing.getHorizontal(meta));
```

gr8pefish replied:
Ah yes, meant to refactor this from the bitwise manipulation I had before with the Variant IProperty; I'll do so.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+    @Override
+    public int getMetaFromState(IBlockState state) {
+        int facing = state.getValue(FACING).ordinal() - EnumFacing.Plane.VERTICAL.facings().length; //0-3, ignore Up and Down
+        return facing; //first 2 bits for facing
```
```java
return state.getValue(FACING).getHorizontalIndex();
```

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+
+    //stolen from vanilla furnace
+    @Override
+    public void onBlockAdded(World worldIn, BlockPos pos, IBlockState state) {
```
Stop this cancer. Use `getStateForPlacement(...)` and base it on the placer's horizontal orientation.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/BlockBackpack.java:
```diff
@@ -0,0 +1,255 @@
+     * Called by ItemBlocks after a block is set in the world, to allow post-place logic
+     */
+    public void onBlockPlacedBy(World worldIn, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack) {
+        worldIn.setBlockState(pos, state.withProperty(FACING, placer.getHorizontalFacing().getOpposite()), 2);
```
boi stop this duplicated placement code please. You're doing the exact same thing *3 times*.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/TileEntityBackpack.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public boolean hasCapability(@Nonnull Capability<?> capability, @Nullable EnumFacing facing) {
+        return capability == IronBackpacksInventoryHelper.BACKPACK_INV_CAPABILITY;
```
Should also react to the normal inventory cap. Also needs to check super for caps attached by the event.

gr8pefish replied:
As in it should provide a normal IItemHandler for external manipulation (i.e. a pipe-able chest)?

Yes, good call on the super.

TehNut replied:
Not specifically for pipes, but yeah.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/TileEntityBackpack.java:
```diff
@@ -0,0 +1,77 @@
+    @Nullable
+    @Override
+    public <T> T getCapability(@Nonnull Capability<T> capability, @Nullable EnumFacing facing) {
+        return capability == IronBackpacksInventoryHelper.BACKPACK_INV_CAPABILITY ? (T) this : null;
```
Need to return `super`, not `null` to check for caps attached by the event.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/block/TileEntityBackpack.java:
```diff
@@ -0,0 +1,77 @@
+
+
+    @Override
+    public NBTTagCompound getTileData() {
```
boi

TehNut commented on src/main/java/gr8pefish/ironbackpacks/container/ContainerBackpack.java:
```diff
@@ -31,16 +33,46 @@
     private final BackpackSize backpackSize;
-    @Nonnull
+    @Nullable
     private final ItemStack backpackStack;
```
No. This is 1.12. Stacks should never be null. Use `ItemStack.EMPTY`.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/container/ContainerBackpack.java:
```diff
@@ -31,16 +33,46 @@
 
-    public ContainerBackpack(@Nonnull ItemStack backpackStack, @Nonnull InventoryPlayer inventoryPlayer, @Nonnull EnumHand hand) {
+    /** Constructor for the TE */
+    public ContainerBackpack(@Nonnull InventoryPlayer inventoryPlayer, @Nonnull EnumHand hand, @Nonnull TileEntityBackpack backpackTileEntity) {
```
I don't really like how this is handled, but I can't really give a better impl right now.

gr8pefish replied:
Yeah, I was thinking of passing in some (IBackpack?) that delegated the methods inside to either the ItemStack or the TE, as I agree, it isn't super nice right now. But I ran into some issues with that and an overloaded constructor (with some non-optimal fields) seemed simpler for the time being.

gr8pefish committed:
Change creative tab name back

gr8pefish committed:
Add default constructor to TE

Also add setter method for BackpackInfo

gr8pefish committed:
BackpackBlock cleanup

Remove unused methods, refactor others.
Mostly changes from TehNut's code review

gr8pefish committed:
TileEntityBackpack cleanup

Same idea as before

gr8pefish committed:
Move methods from API to TEBackpack

Also cleanup them somewhat

gr8pefish committed:
Add block placement code

TehNut commented on src/main/java/gr8pefish/ironbackpacks/item/ItemBackpack.java:
```diff
@@ -70,6 +77,48 @@ public ICapabilityProvider initCapabilities(ItemStack stack, NBTTagCompound oldC
+
+                    BackpackInfo info = getBackpackInfoHeld(player, hand).getKey();
+
+                    if (world.mayPlace(backpackBlock, pos.up(), false, side, player)) {
```
Should be based on the facing you are passed. Don't just assume up.

gr8pefish committed:
Start of Block registration

Change methods to call TEBackpacks' methods, not the outdated BackpackInfo one
Chnage GuiHandler calls to be the right parameter order (why wasn't this right before?)

Overall not working to place blocks (still?)
